### DISCUSSION TITLE
Can you just copy paste the kubeconfig from one master to another?

### DISCUSSION THREAD

**Discussion by revosw** [k3s-io/k3s#5475] [Q&A]
Sorry for rambling and this being an [XY problem](https://xyproblem.info/)

From what I understand, `sudo k3s server` is going to generate new certificates and stuff, but if I want a 3-master HA cluster. Is it possible to copy the kubeconfig or anything related from one master to another as an alternative to joining the cluster without cli commands?

To preface:
- I'm gonna use embedded etcd
- I'm using raspberry pis with SD cards
- I'm installing version 1.22.7+k3s1
- I'm not running any of the commands on my machine - i SSH to a raspberry pi and run everything there
The problem is that when I run this command on the first master...
```sh
curl -sfL https://get.k3s.io | K3S_TOKEN=test123 K3S_INSTALL_EXEC="--cluster-init" sh -
```
...then this on the second master...
```sh
curl -sfL https://get.k3s.io | K3S_TOKEN=test123 K3S_INSTALL_EXEC="--server https://192.168.1.150:6443" sh -
```
then run `kubectl get nodes`, the nodes haven't joined the same cluster.

I've also tried copying the yaml file from `/etc/rancher/k3s/k3s.yaml` to `~/.kube/config`  and `export KUBECONFIG=~/.kube/config` and `sudo chown user:user .kube/config`

It also seems strange to me that no matter what I pass to the `--server` flag, on first time installation, the `clusters[*].cluster.server` key in kubeconfig is always `https://127.0.0.1:6443`. [Shouldn't it point to the other master](https://rancher.com/docs/k3s/latest/en/cluster-access/)? Or is that only for people managing the cluster with kubectl?

I just fail to understand what exactly is the prerequisites for one node to join as a master. 
- if you don't replace the old kubeconfig after k3s reinstallation, the `client-certificate-data` won't match with newly generated certificates and you get a `x509 something something`
- K3S_TOKEN must match the other master
- What else?

**Comment by brandond** [comment]
The kubeconfig only contains enough information for a Kubernetes client to communicate to an apiserver. It does not contain join tokens, cluster configuration, datastore endpoints, or anything else that a new server would need to know when joining the cluster.

* If you're using embedded etcd, you need the address of an existing server and the token.
* If you're using an external SQL datastore, you need the datastore DSN and the token.
* If you're using embedded sqlite, you cannot join additional servers until converting to embedded etcd.

The kubeconfig on servers always points at localhost, because the apiserver is always running locally on servers. Why would it need to talk to another server when it has an apiserver locally?

**Reply by revosw** [reply]
> The kubeconfig on servers always points at localhost, because the apiserver is always running locally on servers. Why would it need to talk to another server when it has an apiserver locally?

That clears up a fundamental misunderstanding I have about kubeconfig. Thanks!

I'll try to post a step-by-step of what I do tomorrow, and maybe in the process I'll find something that works. Really been struggling with joining another master to the cluster

**Reply by brandond** [reply]
It should be dead simple. What do you see, other than just the second node not showing up on the first? 

Have you confirmed that the firewall is disabled on both hosts? Do both of the hosts have unique hostnames?

Also, you really shouldn't use embedded etcd on a Rasbperry Pi; etcd will destroy SD cards. I would probably recommend just doing 1 server (without --cluster-init) and 2 agents, unless you have external SSD or NVME drives for all three of them.
